src/lib/greenboot.rs

```diff
@@ -191,13 +208,102 @@ mod test {
     }
 
     #[test]
-    fn test_failed_diagnostics() {
-        setup_folder_structure(false)
-            .context("Test setup failed")
+    fn test_required_script_failure_exit_early() {
+        init_logger();
+        setup_folder_structure_for_test_required_script_failure_counter().expect("setup failed");
+
+        let base_path = GREENBOOT_INSTALL_PATHS[1];
+        let counter_file = format!("{}/fail_counter.txt", base_path);
+
+        let result = run_diagnostics(vec![]);
+        log::debug!("Diagnostics result: {:?}", result);
+
+        assert!(result.is_err());
+        assert_eq!(
+            result.unwrap_err().to_string(),
+            "health-check failed, skipping remaining scripts"
+        );
+
+        log::info!("Health check failed as expected.");
+
+        let fail_script_count = fs::read_to_string(counter_file)
+            .unwrap()
+            .trim()
+            .parse::<u32>()
             .unwrap();
-        let failed_msg = run_diagnostics(vec![]).unwrap_err().to_string();
-        assert_eq!(failed_msg, String::from("health-check failed!"));
-        tear_down().context("Test teardown failed").unwrap();
+        assert_eq!(
+            fail_script_count, 1,
+            "Only one failing script should have executed"
+        );
+
+        tear_down_setup_exit_early().expect("teardown failed");
+    }
+
+    fn setup_folder_structure_for_test_required_script_failure_counter() -> Result<()> {
+        let base_path = GREENBOOT_INSTALL_PATHS[1];
+        let required_path = format!("{}/check/required.d", base_path);
+        fs::create_dir_all(&required_path)?;
+
+        let passing_script = "testing_assets/passing_script.sh";
+        let failing_script = "testing_assets/failing_script.sh";
+
+        // Counter file for tracking execution
+        let counter_file = format!("{}/fail_counter.txt", base_path);
+        let mut file = File::create(&counter_file)?;
+        writeln!(file, "0")?;
+
+        // add two passing scripts
+        for name in ["00_pass", "01_pass"].iter() {
+            let dest = format!("{}/{}.sh", required_path, name);
+            fs::copy(passing_script, &dest)?;
+            fs::set_permissions(&dest, fs::Permissions::from_mode(0o755))?;
+        }
+
+        // add two failing scripts with counter logic
+        for name in ["10_fail", "20_fail"] {
+            let dest = format!("{}/{}.sh", required_path, name);
+            let original = fs::read_to_string(failing_script)?;
+            let script = format!(
+                "#!/bin/bash\nCOUNTER_FILE=\"{}/fail_counter.txt\"\ncount=$(cat $COUNTER_FILE)\necho $((count + 1)) >| $COUNTER_FILE\n{}",
+                base_path, original
+            );
+            let mut f = File::create(&dest)?;
+            f.write_all(script.as_bytes())?;
+            fs::set_permissions(&dest, fs::Permissions::from_mode(0o755))?;
+        }
+
+        Ok(())
+    }
+
+    fn tear_down_setup_exit_early() -> Result<()> {
```

Comment:
There is already a function `tear_down()` that deletes everything in `GREENBOOT_INSTALL_PATHS[1]`

Reply:
right.. since we are removing everything thats under GREENBOOT_INSTALL_PATHS , we can use common function for all tests. updated the test now.